File: libs/power-log-parser/src/lib/parsers/decklist-update-parser.ts
import { CardType, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { Action, FullEntity, Node } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

const DECK_ID_SCENARIOS = [3428, 3429, 3430, 3431, 3432, 3438, 3433, 3434, 3435, 3436, 3437, 3439];

export class DecklistUpdateParser implements ActionParser {
	readonly ParserName = 'DecklistUpdateParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return stateType === StateType.PowerTaskList && this.isDecklistUpdateAction(node);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const cardsCreatedInDeck = action.Data.filter((data): data is FullEntity => data instanceof FullEntity).filter(
			(entity) => entity.GetTag(GameTag.ZONE) === (Zone.DECK as number),
		);
		if (cardsCreatedInDeck == null || cardsCreatedInDeck.length === 0) {
			return null;
		}

		const timestamp = cardsCreatedInDeck[cardsCreatedInDeck.length - 1].TimeStamp;
		const fullEntities = action.Data.filter((data): data is FullEntity => data instanceof FullEntity)
			.filter((entity) => entity.GetTag(GameTag.ZONE) === (Zone.PLAY as number))
			.filter((entity) => {
				if (DECK_ID_SCENARIOS.includes(this.StateFacade.ScenarioID)) {
					return entity.GetTag(GameTag.HERO_DECK_ID) > 0;
				}
				return (
					entity.GetTag(GameTag.CARDTYPE) === (CardType.HERO as number) &&
					entity.GetEffectiveController() === this.StateFacade.OpponentPlayer!.PlayerId
				);
			});

		return fullEntities.map((fullEntity) => {
			const deckId = DECK_ID_SCENARIOS.includes(this.StateFacade.ScenarioID)
				? '' + fullEntity.GetTag(GameTag.HERO_DECK_ID)
				: fullEntity.CardId;
			const controllerId = fullEntity.GetEffectiveController();
			return GameEventProvider.Create(
				timestamp,
				'DECKLIST_UPDATE',
				GameEventHelper.CreateProvider('DECKLIST_UPDATE', null as any, controllerId, -1, this.StateFacade, {
					DeckId: deckId,
				}),
				true,
				node,
			);
		});
	}

	private isDecklistUpdateAction(node: Node): boolean {
		if (node.Type !== Action) {
			return false;
		}
		const action = node.Object as Action;
		return (
			action.Data != null &&
			action.Data.filter((data): data is FullEntity => data instanceof FullEntity)
				.filter((entity) => entity.GetTag(GameTag.ZONE) === (Zone.PLAY as number))
				.filter((entity) => {
					if (DECK_ID_SCENARIOS.includes(this.StateFacade.ScenarioID)) {
						return entity.GetTag(GameTag.HERO_DECK_ID) > 0;
					}
					return (
						entity.GetTag(GameTag.CARDTYPE) === (CardType.HERO as number) &&
						entity.GetEffectiveController() === this.StateFacade.OpponentPlayer!.PlayerId
					);
				}).length > 0
		);
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/DecklistUpdateParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class DecklistUpdateParser : ActionParser
    {

        private static List<int> DECK_ID_SCENARIOS = new List<int> { 
            // ToT normal and heroic
            3428, 3429, 3430, 3431, 3432, 3438, 3433, 3434, 3435, 3436, 3437, 3439, };

        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public DecklistUpdateParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        // We use the action so that the event is emitted after all the "create card in deck" events
        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && IsDecklistUpdateAction(node);
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Parser.ReplayData.GameActions.Action;
            var cardsCreatedInDeck = action.Data
                    .Where(data => data.GetType() == typeof(FullEntity))
                    .Select(data => (FullEntity)data)
                    .Where(entity => entity.GetTag(GameTag.ZONE) == (int)Zone.DECK)
                    .ToList();
            if (cardsCreatedInDeck == null || cardsCreatedInDeck.Count == 0)
            {
                return null;
            }
            // Time the "new deck" action right after all the "create card in deck" ones
            var timestamp = cardsCreatedInDeck.Last().TimeStamp;
            var fullEntities = action.Data
                    .Where(data => data.GetType() == typeof(FullEntity))
                    .Select(data => (FullEntity)data)
                    .Where(entity => entity.GetTag(GameTag.ZONE) == (int)Zone.PLAY)
                    .Where(entity =>
                     {
                         if (DECK_ID_SCENARIOS.Contains(StateFacade.ScenarioID))
                         {
                             return entity.GetTag(GameTag.HERO_DECK_ID) > 0;

                         }
                         return entity.GetTag(GameTag.CARDTYPE) == (int)CardType.HERO
                             && entity.GetEffectiveController() == StateFacade.OpponentPlayer.PlayerId;
                     })
                    .ToList();
            return fullEntities.Select(fullEntity =>
                {
                    var deckId = DECK_ID_SCENARIOS.Contains(StateFacade.ScenarioID)
                        ? "" + fullEntity.GetTag(GameTag.HERO_DECK_ID)
                        : fullEntity.CardId;
                    var controllerId = fullEntity.GetEffectiveController();
                    return GameEventProvider.Create(
                        timestamp,
                        "DECKLIST_UPDATE",
                        GameEvent.CreateProvider(
                            "DECKLIST_UPDATE",
                            null,
                            controllerId,
                            -1,
                            StateFacade,
                            //null,
                            new
                            {
                                DeckId = deckId,
                            }),
                        true,
                        node);
                })
                .ToList();
        }

        private bool IsDecklistUpdateAction(Node node)
        {
            if (node.Type != typeof(Parser.ReplayData.GameActions.Action))
            {
                return false;
            }
            var action = node.Object as Parser.ReplayData.GameActions.Action;
            return action.Data != null
                && (action.Data
                    .Where(data => data.GetType() == typeof(FullEntity))
                    .Select(data => (FullEntity)data)
                    .Where(entity => entity.GetTag(GameTag.ZONE) == (int)Zone.PLAY)
                    .Where(entity =>
                    {
                        if (DecklistUpdateParser.DECK_ID_SCENARIOS.Contains(StateFacade.ScenarioID))
                        {
                            return entity.GetTag(GameTag.HERO_DECK_ID) > 0;

                        }
                        return entity.GetTag(GameTag.CARDTYPE) == (int)CardType.HERO
                            && entity.GetEffectiveController() == StateFacade.OpponentPlayer.PlayerId;
                    })
                    .ToList()
                    .Count > 0);
        }
    }
}
